WIP: dtoverlays: Update CSI2 sensor overlays to add a cam0 parameter

6by9 opened the pull request:
Add a cam0 parameter to the CSI2 sensor overlays to switch
to the CAM0 connector on Compute Module IO boards.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Update CSI2 sensor overlays to add a cam0 parameter

Add a cam0 parameter to the CSI2 sensor overlays to switch
to the CAM0 connector on Compute Module IO boards.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
@pelwell Could I have your review on this before I continue and modify them all?

One issue I do still have is on the tc358743 overlay I can't get it to add in the data-lanes property based on the ```4lane``` override. I seem to remember you saying you can't amend nodes that are added in the same overlay or something. Could you advise please?

pelwell commented:
Fragments 7 and 8 still refer to csi1_ep, which is undefined. Try with that fixed.

6by9 commented:
> Fragments 7 and 8 still refer to csi1_ep, which is undefined. Try with that fixed.

Doh, too much copy and pasting. I'm sure I had had that fixed and it still didn't want to play, but it is the first thing to check.

6by9 commented:
Updated. Still fails
```
007323.727: dtdebug: Opened overlay file 'overlays/tc358743.dtbo'
007324.862: brfs: File read: /mfs/sd/overlays/tc358743.dtbo
007412.787: Loaded overlay 'tc358743'
007412.812: dtparam: 4lane=true
007413.776: dtdebug: found override 4lane
007422.650: dtdebug: merge_fragment(/fragment@100/__overlay__,/fragment@0/__overlay__)
007422.695: dtdebug:   +prop(#address-cells)
007422.837: dtdebug:   +prop(#size-cells)
007422.992: dtdebug:   +prop(status)
007424.435: dtdebug: merge_fragment(/fragment@100/__overlay__/tc358743@0f,/fragment@0/__overlay__/tc358743@0f)
007424.481: dtdebug:   +prop(compatible)
007424.597: dtdebug:   +prop(reg)
007424.732: dtdebug:   +prop(status)
007424.877: dtdebug:   +prop(clocks)
007425.035: dtdebug:   +prop(clock-names)
007426.617: dtdebug: merge_fragment(/fragment@100/__overlay__/tc358743@0f/port,/fragment@0/__overlay__/tc358743@0f/port)
007428.035: dtdebug: merge_fragment(/fragment@100/__overlay__/tc358743@0f/port/endpoint,/fragment@0/__overlay__/tc358743@0f/port/endpoint)
007428.080: dtdebug:   +prop(remote-endpoint)
007428.206: dtdebug:   +prop(clock-lanes)
007428.343: dtdebug:   +prop(clock-noncontinuous)
007428.499: dtdebug:   +prop(link-frequencies)
007428.666: dtdebug:   +prop(phandle)
007428.864: dtdebug: merge_fragment() end
007428.915: dtdebug: merge_fragment() end
007428.980: dtdebug: merge_fragment() end
007429.080: dtdebug: merge_fragment() end
007433.741: dtdebug: merge_fragment(/fragment@101/__overlay__,/fragment@1/__overlay__)
007433.786: dtdebug:   +prop(status)
007435.536: dtdebug: merge_fragment(/fragment@101/__overlay__/port,/fragment@1/__overlay__/port)
007437.273: dtdebug: merge_fragment(/fragment@101/__overlay__/port/endpoint,/fragment@1/__overlay__/port/endpoint)
007437.317: dtdebug:   +prop(remote-endpoint)
007437.438: dtdebug:   +prop(data-lanes)
007437.580: dtdebug:   +prop(phandle)
007437.745: dtdebug: merge_fragment() end
007437.789: dtdebug: merge_fragment() end
007437.847: dtdebug: merge_fragment() end
007438.395: dtdebug: fragment 2 disabled
007439.870: dtdebug: merge_fragment(/fragment@0/__dormant__/tc358743@0f/port/endpoint,/fragment@3/__overlay__)
007439.914: dtdebug:   +prop(data-lanes)
007440.160: dtdebug: merge_fragment() end
007449.727: dtdebug: fragment 7 disabled
007452.057: dtdebug: merge_fragment(/fragment@1/__dormant__/port/endpoint,/fragment@8/__overlay__)
007452.103: dtdebug:   +prop(data-lanes)
007452.270: dtdebug: merge_fragment() end
007462.888: dtdebug: fragment 0 disabled
007463.200: dtdebug: fragment 1 disabled
007463.419: dtdebug: fragment 2 disabled
007463.595: dtdebug: fragment 3 disabled
007491.523: dtdebug: merge_fragment(/soc/i2c@7e205000,/fragment@4/__overlay__)
007491.566: dtdebug:   +prop(status)
007492.984: dtdebug: merge_fragment() end
007521.137: dtdebug: merge_fragment(/soc/i2c0mux,/fragment@5/__overlay__)
007521.183: dtdebug:   +prop(status)
007522.500: dtdebug: merge_fragment() end
007523.361: dtdebug: merge_fragment(/,/fragment@6/__overlay__)
007549.820: dtdebug: merge_fragment(/bridge-1-clk,/fragment@6/__overlay__/bridge-1-clk)
007549.867: dtdebug:   +prop(compatible)
007551.747: dtdebug:   +prop(#clock-cells)
007553.720: dtdebug:   +prop(clock-frequency)
007555.683: dtdebug:   +prop(phandle)
007557.688: dtdebug: merge_fragment() end
007557.739: dtdebug: merge_fragment() end
007557.955: dtdebug: fragment 7 disabled
007558.133: dtdebug: fragment 8 disabled
007587.145: dtdebug: merge_fragment(/soc/i2c0mux/i2c@1,/fragment@100/__overlay__)
007587.187: dtdebug:   +prop(#size-cells)
007588.467: dtdebug:   +prop(#address-cells)
007589.708: dtdebug:   +prop(status)
007602.206: dtdebug: merge_fragment(/soc/i2c0mux/i2c@1/tc358743@0f,/fragment@100/__overlay__/tc358743@0f)
007602.251: dtdebug:   +prop(clock-names)
007603.618: dtdebug:   +prop(clocks)
007604.938: dtdebug:   +prop(status)
007606.268: dtdebug:   +prop(reg)
007607.628: dtdebug:   +prop(compatible)
007620.132: dtdebug: merge_fragment(/soc/i2c0mux/i2c@1/tc358743@0f/port,/fragment@100/__overlay__/tc358743@0f/port)
007631.369: dtdebug: merge_fragment(/soc/i2c0mux/i2c@1/tc358743@0f/port/endpoint,/fragment@100/__overlay__/tc358743@0f/port/endpoint)
007631.409: dtdebug:   +prop(phandle)
007632.704: dtdebug:   +prop(link-frequencies)
007634.637: dtdebug:   +prop(clock-noncontinuous)
007636.579: dtdebug:   +prop(clock-lanes)
007638.552: dtdebug:   +prop(remote-endpoint)
007640.527: dtdebug: merge_fragment() end
007640.579: dtdebug: merge_fragment() end
007640.644: dtdebug: merge_fragment() end
007640.745: dtdebug: merge_fragment() end
007682.560: dtdebug: merge_fragment(/soc/csi@7e801000,/fragment@101/__overlay__)
007682.605: dtdebug:   +prop(status)
007698.319: dtdebug: merge_fragment(/soc/csi@7e801000/port,/fragment@101/__overlay__/port)
007713.031: dtdebug: merge_fragment(/soc/csi@7e801000/port/endpoint,/fragment@101/__overlay__/port/endpoint)
007713.075: dtdebug:   +prop(phandle)
007713.959: dtdebug:   +prop(data-lanes)
007714.967: dtdebug:   +prop(remote-endpoint)
007716.475: dtdebug: merge_fragment() end
007716.520: dtdebug: merge_fragment() end
007716.579: dtdebug: merge_fragment() end
```
/proc/device-tree/soc/i2c0mux/i2c@1/tc358743@0f/port/endpoint/ shows no data-lanes property (fragment 2 or 3), but /proc/device-tree/soc/csi@7e801000/port/endpoint/  (fragment 7or 8) does.

I thought I'd sorted these overlays so that they could be loaded with ```sudo dtoverlay foo```, but I'm seeing errors getting clocks again, and
```
OF: /soc/i2c0mux/i2c@1/tc358743@0f: could not get #clock-cells for /soc/avs-monitor@7d5d2000/thermal
```
which seems weird.

pelwell commented:
There are limitations to the reordering done by the fragment application code - it will apply all fragments that target other fragments before those that target the base dtb, but that's as far as it goes. The safest way to write an overlay is to list the fragments in the order in which they should be applied.

If you move fragments 2 & 3 to be fragments 0 & 1, with the old fragment 0 going to 2, then 7 & 8 -> 3 & 4, 1 -> 5,  4->6, 5->7 and 6->8, remembering to renumber the 4lane parameter, then it will work as expected.

This has been a useful exercise because it highlighted a bug in ovmerge - now fixed.

6by9 commented:
Thanks for the advice - this is why I consult you as the DT expert :-)

TC358743 updated to allow for setting the number of lanes, and that is working as intended.
It still won't load via ```dtoverlay```, but this time because the driver is told the clock is 54MHz and not 27MHz. Most odd as /sys/kernel/debug/clk/clk_summary reports the clock as 27MHz. I'm not going to dwell on that one now unless you immediately know the answer.

If you're happy with the styling of these overlays then I'll go ahead and update the others.

pelwell commented:
Yes, I'm happy with those. I should try to update the utilities (dtoverlay, dtmerge, ovmerge) to at least warn when it modifies a fragment that has already been merged.

6by9 committed:
media/ov7251: Make Enable GPIO optional.

As with many sensors, external control on most OV7251 modules is
via a common GPIO that controls the regulators, clock, and sensor
shutdown/reset line.
Handling that line via a regulator driver rather than within the
sensor driver allows the same GPIO to be shared by multiple sensors,
so switch to that behaviour.

Commit 79fc876cf183 "dtoverlays: Update sensor overlays to use
cam1_reg where possible" had actually broken this driver as it
switched to a common regulator and had removed the enable GPIO.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented:
Is this ready to merge yet? It's still marked as WIP.

6by9 commented:
I hit further problems over the updating of the regulator. It makes no difference on CM4 as the two CAM ports share an enable GPIO, but it does make a difference on CM1/3 as generally they are different.

pelwell commented:
OK - I'll await further instructions.

Does using multiple labels help, e.g. cam0_reg and cam1_reg, where they are both on the same node in the CM4 case? I hope reference-counting is necessary...

6by9 commented:
I suspect that it's the ordering thing that we keep discussing - most likely it was updating a fragment that had already been applied.
For CM4 it doesn't matter if it still actually points at cam1_reg as it's the same node. For CM1/3 it does matter.

The regulator framework does all the reference counting for you. 
```sudo cat /sys/kernel/debug/regulator/regulator_summary``` will tell you the use counts for each reg, and it only powers down when all users have gone (and it'll delay a smidge after that too in case you really wanted it again almost immediately).

pelwell commented:
I'm happy to review anything if you think it might help.

6by9 committed:
WIP: dtoverlays: Update the regulator as well as the CSI and I2C nodes

6by9 commented:
Branch updated with imx477 reworked to also update the regulator node. I'm about to rebuild it for CM3 again. It works on CM4, but I'm fairly certain that it doesn't update the regulator node correctly.

6by9 commented:
Now I'm puzzled as this has just worked ok on a CM3, but I think I may have killed GPIOs 0&1 now.

Minor annoyance is that the wiring is not the same as for the firmware. The docs say to connect the camera on CAM1 to GPIOs 0&1, which would be the normal i2c0 rather than i2c_csi_dsi.

Second CM3 and CMIO also not showing sensor on GPIOs 0&1. More investigation required.

pelwell commented on arch/arm/boot/dts/overlays/ov5647-overlay.dts:
```diff
@@ -76,17 +76,49 @@
 		rotation = <&ov5647>,"rotation:0";
+		cam0 =	<&frag100>, "target:0=",<&i2c_vc>,
+			<&frag101>, "target:0=",<&csi0>,
+			<&frag102>, "target:0=",<&cam0_reg>,
```
There is no label `frag102` in this overlay.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -1631,6 +1631,8 @@ Info:   Sony IMX219 camera module.
 Load:   dtoverlay=imx219,<param>=<val>
 Params: rotation                Mounting rotation of the camera sensor (0 or
                                 180, default 180)
+        cam0                    Adopt the default configuration for CAM0 on a
```
The `adv7282m` and `adv7282x-m` overlays have also gained a `cam0` parameter.

pelwell reviewed (commented):
Some observations at this stage.

pelwell commented:
Superseded by #4727.
